src/builders.rs

```diff
@@ -1208,3 +1240,20 @@ impl fmt::Debug for Output {
         self.0.fmt(f)
     }
 }
+
+#[derive(Debug)]
+pub struct CustomFileLoggerInfo{
```

Comment:
In the interest of keeping names concise and descriptive, what would you think of renaming this to `DateBasedFile` or something similar?

Ideally I'd like to avoid the `Info` suffix, since it doesn't tell us anything. Everything in this module is configuration info. The user doesn't need to know or specify that they're talking about Info since there's no non-info structs in Fern they can access.

Then `Custom` also doesn't say much, but it's not bad. For usability, naming something unique based on its capabilities makes it easier to find, though. The unique thing about this logger is that it varies file name in time, which is why I was thinking `DateBased` or `TimeBased` would make sense.

Reply:
I changed the external facing api to have following functions
date_based_log_file
hour_based_log_file
minute_based_log_file

let me know what you think,
i was thinking that the same structure could be enhanced to support size based log file rotation also in the future.

Reply:
Alright - names are good! I think they could be even shorter & cleaner as static methods on `CustomFileLoggerInfo` though.

In my initial review, I missed something and thought this structure _was_ part of the public API. It's not re-exported from `lib.rs`, though, so it isn't. For usability and consistency with the rest of fern, I think it probably should be part of the public API.

Right now, most of the configuration structures are public - `Output`, etc. are accessible and provide static methods rather than freestanding functions to manipulate them. Two reasons for these two strategies:

1. Having all structures returned as part of the public API public themselves ensuring the user can always name all structures that they have access to.

    When we return a structure that isn't exposed to the outside of the crate, the user can call methods on it, and use it from things, but they can't access its documentation, and can't return it from a method because they have no name for what to call it.
    
    I consider this a bad practice just because it's giving the user things to work with that they can't _really_ work with. With the PR as it is right now, someone can't write

    ```rust
    fn configure_date() -> CustomFileLoggerInfo {
        fern::date_based_log_file("my-log.log")
    }
    ```

    because they can't import `fern::CustomFileLoggerInfo`.

    In addition, they can't look at the documentation for `CustomFileLoggerInfo` to see that it implements `Into<Output>`. They can guess that it'd be valid to pass into `Dispatch::chain`, but that fact wouldn't be confirmable from the docs.

2. Using static methods rather than freestanding functions namespaces things in a cleaner fashion.

    Rather than having 100% of the public API need to be in the same `fern::*` namespace as functions, having all the constructors and methods under a structure allows the user to only consider part of the API at the time. Kind of modularizing it, kind of just keeping it organized.

    For example, we have `Dispatch::new()` rather than `fern::new_dispatch`, then `Dispatch::apply()` rather than `fern::set_logger(final)`. The one exception to this is in the current API is `fern::log_file()`, but _the only reason for that_ is that it returns an `std::io::File`, a structure we don't control. It's strictly utility, and not really related to anything else in the API.

    I'd also like to think this also helps discoverability. Having all the different things as methods on a structure allows looking through a short list of things that one can do with just that object, and thus actually find those methods. If I had spent more time designing it better, `fern::log_file` would probably be a method on `Output` for this reason.

    Unlike the situation with `log_file`, though, these structures directly return a structure we define, and thus could directly belong to that structure's namespace as static methods.

---

I think supporting size-based loggers is a good idea. We can definitely share some backend code for this, too. But on the front end, why not have a different structure?

"Appending the date" and "rotating when size gets too big" are pretty distinct from an API perspective, especially since "writing to a file appending the date" and "writing to a file" are already separate.

We need different configuration options for each - in one case, we need a prefix and time-based suffix, and in the other, we need a file name and strategy for finding new file names (or moving old files) when size overflows. I'm making some assumptions here about the size-rotating file API, but regardless it seems like it'd at least store some different configuration information from the time-based one.